### DISCUSSION TITLE
How to build xmake from source?

### DISCUSSION THREAD

**Discussion by f-zl** [xmake-io/xmake#3934] [Q&A]
Hello, thanks for the great tool. I have some trouble building xmake from source. I'm building xmake on Windows, running `./configure` and `make` in a msys2 mingw64 shell. The target `build/xmake.exe` can be built. But running it always fails with nothing printed. How can I build xmake correctly?

Result of `./configure`:
> checking for platform ... mingw
checking for architecture ... x86_64
checking for make ... ok
checking for the c compiler (cc) ... x86_64-w64-mingw32-gcc
checking for the c++ compiler (cxx) ... x86_64-w64-mingw32-gcc
checking for the assembler (as) ... x86_64-w64-mingw32-gcc
checking for the objc compiler (mm) ... x86_64-w64-mingw32-gcc
checking for the objc++ compiler (mxx) ... x86_64-w64-mingw32-gcc
checking for the linker (ld) ... x86_64-w64-mingw32-g++
checking for the static library archiver (ar) ... ar
checking for the shared library linker (sh) ... x86_64-w64-mingw32-g++
checking for toolchain ... x86_64_w64_mingw32
checking for readline .. no
checking for curses .. no
checking for atomic .. ok
checking for lua .. ok
checking for luajit .. no
checking for lz4 .. ok
checking for sv .. no
checking for tbox .. no
checking for wchar .. no

Generated xmake.config.h:
```
#ifndef XM_CONFIG_H
#define XM_CONFIG_H

// version
#define XM_CONFIG_VERSION "2.7.9"
#define XM_CONFIG_VERSION_MAJOR 2
#define XM_CONFIG_VERSION_MINOR 7
#define XM_CONFIG_VERSION_ALTER 9
#define XM_CONFIG_VERSION_BUILD 20230707
#define XM_CONFIG_VERSION_BRANCH "master"
#define XM_CONFIG_VERSION_COMMIT "c801c64e8"

#endif
```

**Comment by waruqi** [comment]
please use xmake to build xmake on windows. 

https://xmake.io/#/guide/installation?id=source-compilation-in-windows-platform

**Reply by f-zl** [reply]
Thanks. I fail to build xmake using xmake, although the prebuilt xmake works properly on other projects. Using mingw, the compilation fails at lua(lauxlib.c) because there is no <sys/wait.h>. Using msvc, the built xmake.exe always prints these lines and does not function:
> [xmake]: init: ..
[xmake]: version: xmake_v2_7_9_20230707_x64 by unknown visual c++ compiler
[tbox]: init: ..
[tbox]: [dns]: server: 192.168.137.1
[tbox]: version: tbox_v1_7_3_20230707_x64 by unknown visual c++ compiler
[tbox]: init: ok
[xmake]: init: ok
[tbox]: environment variable(MSYSTEM) does not exist
[tbox]: environment variable(MSYSTEM_CARCH) does not exist
[tbox]: environment variable(XMAKE_PROFILE) does not exist

Build of xmake in wsl succeeds. I'll use the wsl build for debugging.

**Comment by waruqi** [comment]
> Using msvc, the built xmake.exe always prints these lines and does not function:

you need export `XMAKE_PROGRAM_DIR` to bind lua scripts, or click srcenv.bat 

https://xmake.io/#/guide/faq?id=loading-a-local-debugging-environment-on-other-platforms

**Comment by f-zl** [comment]
Hi, I want to build xmake from source in a wsl distribution for debugging. The problem is built-from-source xmake prints these lines and exits:
> [xmake]: init: ..
[xmake]: version: xmake_v2_8_6_20240113_x86_64_sse2_elf by gnu c/c++ 13.2

I build it with these commands:
```sh
./configure --mode=debug
make
```

Debugging the xmake executable, I find function `tb_check_mode` in tbox.c fails, the error message in line 90 is `"tbox/tbox.h has __tb_debug__ but libtbox.a not"`. `tb_trace_e` is defined to empty and doesn't print this message.

The flags in the generated makefile:
```makefile
tbox_cflags=  -g  -O0  -Wall  -std=c99  -D_GNU_SOURCE=1 -D_FILE_OFFSET_BITS=64 -D_LARGEFILE_SOURCE -D__tb_prefix__=\"tbox\"  -Icore/src/tbox/tbox/src/tbox/.. -Ibuild/linux/x86_64/debug -Icore/src/tbox/inc/linux
# ...
xmake_cflags=  -g  -O0  -Wall  -std=c99  -D_GNU_SOURCE=1 -D_FILE_OFFSET_BITS=64 -D_LARGEFILE_SOURCE -DXM_CONFIG_API_HAVE_CURSES -D__tb_prefix__=\"xmake\" -D__tb_debug__ -DXM_CONFIG_API_HAVE_LUA_CJSON -DLUA_COMPAT_5_1 -DLUA_COMPAT_5_2 -DLUA_COMPAT_5_3  -Icore/src/xmake/.. -Ibuild/linux/x86_64/debug -Icore/src/xmake/../xxhash -Icore/src/lz4/lz4/lib -Icore/src/sv/sv/include -Icore/src/tbox/tbox/src/tbox/.. -Ibuild/linux/x86_64/debug -Icore/src/tbox/inc/linux -Icore/src/lua/lua -D_DEFAULT_SOURCE -D_XOPEN_SOURCE=600
```

`__tb_debug__` not defined in tbox_cflags . I suppose I shouldn't edit the generated makefile by hand. How can I solve this problem?

**Reply by waruqi** [reply]
try it again on dev branch.
